### PR TITLE
ENG-914: parallel chunked tool calls + adaptive parallelism & sizing (frontend parity)

### PR DESCRIPTION
**Pull request by faiz9068**
## Problem

Production traces of the `last9-mcp` service showed chunked tool calls (`get_logs`, `get_traces`, `get_service_logs`) running for **30–463 seconds** while typical AI/LLM clients time out around 60s. Clients gave up, agents received partial or empty responses, and customers reported hallucinated answers.

Root cause: each chunked tool sliced the time range into **fixed 5-minute windows** and ran them **strictly serially** with **no per-chunk HTTP deadline**. A 2-hour query meant 24 sequential upstream calls; a single slow chunk stalled the whole chain.

## What this PR does

Replaces serial 5-minute chunking with parallel adaptive chunking that mirrors the dashboard's `getAdaptiveLoadingConfig` + `getVolumeQueryChunks` rule-for-rule.

**Parallel fetch with adaptive concurrency** — `internal/utils/parallel_chunks.go` introduces `RunChunksParallel[T]`, a fixed-size-semaphore helper with context-aware acquire so queued goroutines bail on `ctx.Done()`. `internal/utils/adaptive_chunks.go` resolves an `AdaptiveLoadingConfig` per call based on (time range, query shape), driving parallelism between 1 and 6 depending on which of the five rules fire.

**Adaptive chunk sizing** — chunks are no longer fixed-size. `ceil(range/6)` for normal queries, fixed 1-hour chunks when the query has expensive body parsing combined with line-filter optimization. Capped at `MaxChunkSizeMs = 1 h` in all cases. Ranges ≤ 1 h are not split at all.

**Per-chunk HTTP deadline** — `internal/constants/api.go` adds `PerChunkHTTPTimeout = 30 s`. Each chunk derives its own `context.WithTimeout`, so a single slow upstream call can no longer eat the whole budget. The global `DefaultHTTPTimeout = 3 m` stays as an outer envelope.

**Structured error logging** — `slog.Error` on every chunk-error / parse-error path with `tool`, `chunk_index`, `total_chunks`, `start_ms`, `end_ms`, `err` attributes. The server is no longer silent on failures; oncall can `grep tool=get_logs` in the OTLP-routed log stream.

**Pipeline inspection** — `internal/utils/pipeline_inspect.go` ports `hasJSONParsePipeline`, `hasBodyParseOrTransform`, and `hasExpensiveBodyParsing` from the dashboard. `BODY_TRANSFORM_OPERATIONS` matches the frontend list exactly (`$split_into`, `$split`, `$replace_regex`). The bloom-eligibility heuristic is simplified to "any indexable operator on Body (`$eq` / `$ieq`)" because the dashboard's token-count bloom rules are tied to its builder UI.

Callers (`fetchLogJSONQuery`, `fetchTraceJSONQuery`, `fetchServiceLogs`) all build an `AdaptiveLoadingConfig` from their pipeline + time range, then drive `RunChunksParallel` and `GetAdaptiveChunks` with the resolved values.

## Behavior change to note

The old serial loop returned a hard error if the *newest* chunk failed before any results were collected. With parallel execution, all chunks fire concurrently; if any single chunk fails but at least one other succeeds, the response is now annotated as a **partial result** (`_last9_mcp.partial_result: true`, `warning` carries the chunk failure) instead of hard-erroring. This is more resilient and matches the partial-result contract the response already advertises. `fetchLogJSONQuery` and `fetchServiceLogs` use identical semantics post-PR: hard-error only when every chunk failed.

## Measured impact (live smoke against a real backend)

| Workload | Before | After (server wall-clock) |
|---|---|---|
| `get_service_logs` 2 h lookback, limit=100 | 111 s | **< 1 s** |
| `get_service_logs` 12 h lookback, limit=5000 (forces fan-out) | ~360 s+ | **3 s** |
| `get_logs` 2 h, body filter with zero matches (forces full walk) | ~12 s | **1 s** |
| `get_service_logs` 3-day lookback, limit=200 (hits >2 d throttle rule) | ~7 m | **27 s** |

All four cases verified end-to-end via the `LAST9_DEBUG_CHUNKING=true` server stderr, showing the correct rule firing with the correct chunk count and parallelism, and `partial=false` on every run.

## Out of scope

- `addParamsToSpan` typed-struct cast bug — lives in the external `mcp-go-sdk` repo, separate PR.
- Per-org concurrency limit (thundering herd). Important follow-up since this PR multiplies per-call backend load by up to 6×.
- `service.version` and tenant attributes on OTel resource / spans.
- Broader error-handling sweep across non-chunked tools.
- Per-org adaptive overrides — deliberately not ported; MCP server has no settings fetcher today.
- Configurable per-chunk timeout — production traces show individual upstream chunks legitimately taking 25–41 s on cold rehydrated indices. The fixed 30 s default is sensible for hot data but may need an env var for rehydration-heavy workloads. File a follow-up if it bites.

## Test plan

- [x] `go test ./internal/utils/... ./internal/telemetry/logs/... ./internal/telemetry/traces/... -race -count=1` — green
- [x] `go vet ./...` — clean
- [x] `go build ./...` — clean
- [x] Unit coverage for every adaptive rule (small, >2 d, >7 d, >14 d, body+>1 d, line-filter-optimization +/- JSON parse), `RunChunksParallel` order/cap/error/ctx-cancel/timeout, `HasExpensiveBodyParsing` parity edge cases including mixed indexable+non-indexable operators in one stage
- [x] Integration tests for all three handlers — partial-failure path, semaphore cap on 10 h fan-out (≤ `ParallelCallsLimit = 6`)
- [x] Regression test for the `fetchServiceLogs` partial-result fix (5 success-empty chunks + 1 failed chunk no longer hard-errors)
- [x] Live smoke against real backend — all four scenarios in the impact table above confirmed via server stderr
- [ ] Watch upstream backend P95 / connection-pool metrics in the first 24 h after deploy — 6× concurrency multiplier could surface as backend pressure if many agents fire simultaneously

## Summary by CodeRabbit

* **New Features**
  * Queries now execute chunks in parallel for improved performance on large time ranges.
  * Adaptive chunk sizing automatically optimizes based on query characteristics and pipeline complexity.
  * Partial results now returned when some chunks fail, improving resilience.

* **Bug Fixes**
  * Per-chunk timeout prevents single slow requests from stalling entire query execution.

* **Tests**
  * Expanded test coverage for parallel execution, adaptive chunking, and error scenarios.

[Review Change Stack](https://app.coderabbit.ai/change-stack/last9/last9-mcp-server/pull/151?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by faiz9068** [comment]
@coderabbitai full review

File: internal/utils/parallel_chunks.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+			value, err := fn(ctx, i, chunk)
+			results[i] = ChunkResult[T]{
+				Index: i,
+				Chunk: chunk,
+				Value: value,
+				Err:   err,
+			}
+		}(i, chunk)
+	}
```
[the remark this answers is not in the document]
Comment (faiz9068) [reply internal/utils/parallel_chunks.go:66]: Thanks for flagging — pushing back on this one with data.

**The "hundreds/thousands of parked goroutines" claim doesn't match our actual chunk counts.** Adaptive sizing caps chunks at `MaxChunkSizeMs = 1 hour` via `range/6` (see `adaptive_chunks.go`). Worst case in real testing:

| Range | Adaptive chunks |
|---|---|
| 6h | 6 |
| 24h | 24 |
| 3 days | 72 |
| 14 days | 336 |

72 parked goroutines × ~8 KB stack = ~576 KB per request. 336 = ~2.7 MB. Both are well within budget for an HTTP server handling LLM-agent workloads.

**Per-request goroutine cost is dominated by the actual upstream HTTP I/O** (100 ms – 30 s per chunk), not by parked goroutine overhead. The semaphore admission cost is negligible compared to network latency.

**The existing codebase already uses bare `go func()`** in `internal/apm/databases.go` (lines 110-145, five concurrent Prom queries) without a worker pool — same pattern, accepted convention here.

**The proposed worker-pool refactor adds real cost**:
- New `jobs` channel + worker loop adds ~25 lines of indirection.
- Loses the clean per-chunk `ctx.Done()` short-circuit (currently each goroutine checks at semaphore acquire; in the worker-pool variant, jobs already enqueued can't be cancelled without extra plumbing).
- Tests around `TestRunChunksParallelCancelsQueuedChunksOnContextCancel` and `TestRunChunksParallelRespectsConcurrencyCap` would need rework.

**If the load profile changes** (e.g. customers start regularly running 14d+ windows that produce 300+ chunks), this is a worthwhile follow-up. Until then, the current pattern is simple, well-tested, and the per-request memory cost is bounded.

Leaving as-is. Open to revisiting if you see telemetry suggesting goroutine count is actually a bottleneck.

File: internal/telemetry/traces/chunking_test.go [hunk 2]
```diff
@@ -51,19 +55,21 @@ func traceAPIResponse(n int) string {
...
+		// 90-min range → 6 chunks of 15 min each (range/6, no body parse).
+		switch q.Get("start") + "-" + q.Get("end") {
+		case "4500-5400":
 			w.Write([]byte(traceAPIResponse(2)))
-		case 2:
-			// older chunk — returns 2 more traces (only 1 should be kept due to limit=3)
+		case "3600-4500", "2700-3600", "1800-2700", "900-1800":
+			w.Write([]byte(traceAPIResponse(0)))
+		case "0-900":
 			w.Write([]byte(traceAPIResponse(2)))
```
[the remark this answers is not in the document]
Comment (faiz9068) [reply internal/telemetry/traces/chunking_test.go]: Fixed in commit `3c89b08` — all 12 `w.Write` calls in this file now use the `_, _ =` discard form, matching the convention already used in `internal/telemetry/logs/chunking_test.go`. Linter should be clean now.

**Review by prathamesh-sonpatki — APPROVED** [review]
